locator/topology.cc

```diff
@@ -447,11 +445,17 @@ node* topology::find_node(host_id id) noexcept {
 const node* topology::find_node(const inet_address& ep) const noexcept {
     auto it = _nodes_by_endpoint.find(ep);
     if (it != _nodes_by_endpoint.end()) {
-        return it->second;
+        return &it->second.get();
     }
     return nullptr;
 }
 
+// Finds a node by its endpoint
+// Returns nullptr if not found
+node* topology::find_node(const inet_address& ep) noexcept {
+    return make_mutable(const_cast<const topology*>(this)->find_node(ep));
+}
+
```

Comment:
I see that this function is used only in `locator_topology_test.cc`, in lines like `BOOST_REQUIRE_EQUAL(topo.find_node(ep2), node);` to equalize the types. Adding a function that is used only in tests is usually not the best idea, but since it's so simple and we already have the host ID-based version of this function, I guess it's fine.

It took me some time to understand why you added this function because the code you changed doesn't use it. Not obvious changes like this should be explained in the commit message.

BTW, regarding the commit message, you can update it a bit. This PR is not only about `for_each_node` and `const node*` anymore. You also replace `node*` with `node&` in some places. You could just write that we replace `node*` with `node&` in places where `nullptr` is impossible. You can remove `const`s from the title too.

Reply:
Why do we need a mutable method?  Does `find_node` ever modify the topology object?

Reply:
Oh, you need it to return a `node*` instead of a `const node*`?
If so, the conversion should be done at the test.
Adding such a method opens a dangerous backdoor to modifying the `node`.

Reply:
Removed the `find_node(const inet_address& ep)`